pkg/util/k8s/k8s.go

```diff
@@ -0,0 +1,591 @@
+// else updates it if it has changed
+func CreateOrUpdateServiceAccount(
+	k8sClient client.Client,
+	sa *v1.ServiceAccount,
+	ownerRef *metav1.OwnerReference,
+) error {
+	existingSA := &v1.ServiceAccount{}
+	err := k8sClient.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      sa.Name,
+			Namespace: sa.Namespace,
+		},
+		existingSA,
+	)
+	if errors.IsNotFound(err) {
+		logrus.Debugf("Creating %s service account", sa.Name)
+		return k8sClient.Create(context.TODO(), sa)
+	} else if err != nil {
+		return err
+	}
+
+	modified := !reflect.DeepEqual(sa.ImagePullSecrets, existingSA.ImagePullSecrets)
+
+	for _, o := range existingSA.OwnerReferences {
+		if o.UID != ownerRef.UID {
+			sa.OwnerReferences = append(sa.OwnerReferences, o)
+		}
+	}
+
+	if modified || len(sa.OwnerReferences) > len(existingSA.OwnerReferences) {
+		logrus.Debugf("Updating %s service account", sa.Name)
+		return k8sClient.Update(context.TODO(), sa)
+	}
+	return nil
+}
+
+// DeleteServiceAccount deletes a service account if present
+func DeleteServiceAccount(
+	k8sClient client.Client,
+	name, namespace string,
+) error {
+	resource := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+
+	serviceAccount := &v1.ServiceAccount{}
+	err := k8sClient.Get(context.TODO(), resource, serviceAccount)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logrus.Debugf("Deleting %s service account", name)
+	return k8sClient.Delete(context.TODO(), serviceAccount)
+}
+
+// CreateOrUpdateRole creates a role if not present,
+// else updates it if it has changed
+func CreateOrUpdateRole(
+	k8sClient client.Client,
+	role *rbacv1.Role,
+	ownerRef *metav1.OwnerReference,
+) error {
+	existingRole := &rbacv1.Role{}
+	err := k8sClient.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      role.Name,
+			Namespace: role.Namespace,
+		},
+		existingRole,
+	)
+	if errors.IsNotFound(err) {
+		logrus.Debugf("Creating %s role", role.Name)
+		return k8sClient.Create(context.TODO(), role)
+	} else if err != nil {
+		return err
+	}
+
+	modified := !reflect.DeepEqual(role.Rules, existingRole.Rules)
+
+	for _, o := range existingRole.OwnerReferences {
+		if o.UID != ownerRef.UID {
+			role.OwnerReferences = append(role.OwnerReferences, o)
+		}
+	}
+
+	if modified || len(role.OwnerReferences) > len(existingRole.OwnerReferences) {
+		logrus.Debugf("Updating %s role", role.Name)
+		return k8sClient.Update(context.TODO(), role)
+	}
+	return nil
+}
+
+// DeleteRole deletes a role if present
+func DeleteRole(
+	k8sClient client.Client,
+	name, namespace string,
+) error {
+	resource := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+
+	role := &rbacv1.Role{}
+	err := k8sClient.Get(context.TODO(), resource, role)
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	logrus.Debugf("Deleting %s role", name)
+	return k8sClient.Delete(context.TODO(), role)
```

Comment:
Doesn't Delete return IsNotFound error if the object does not exist ?

Reply:
We are doing a get before we call delete above.